crates/chain-state/src/chain_info.rs

```diff
@@ -126,8 +124,11 @@ impl ChainInfoTracker {
     pub fn set_finalized(&self, header: SealedHeader) {
         self.inner.finalized_block.send_if_modified(|current_header| {
             if current_header.as_ref().map(SealedHeader::hash) != Some(header.hash()) {
-                let _ = current_header.replace(header);
-                return true
+                // Only update if the new header has a higher block number
+                if current_header.as_ref().map(|h| h.number) < Some(header.number) {
+                    let _ = current_header.replace(header);
+                    return true;
+                }
```

Comment:
Maybe I'm wrong here but as finalised blocks are unlikely to be reorged, except under very extreme conditions, I think this is an additional security measure to check that the new header has a higher block number than the previous one before modifying the current finalized block header.

But again please do not merge before checking this assumption carefully :)